File: FC.Shared/Entities/MediaDirectory.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FC.Shared.Entities
{
    public class MediaDirectory
    {
        public MediaDirectory()
        {
            //Media = new List<Entities.Media>();
        }

        [Key]
        public Guid? DirectoryID { get; set; }

        public Guid? ParentID { get; set; }
        
        public string Name { get; set; }

        public bool IsDeleted { get; set; }
        [Column(TypeName = "datetime2")]
        public DateTime? ArchiveDate { get; set; }

        public List<Media> Media { get; set; }
        [Column(TypeName = "datetime2")]
        public DateTime? Created { get; set; }
        [Column(TypeName = "datetime2")]
        public DateTime? Modified { get; set; }

        public Guid? AuthorID { get; set; }

        [System.ComponentModel.DataAnnotations.Schema.NotMapped]
        public List<MediaDirectory> Children { get; set; }
    }
}

File: FC.WebMVC/Scripts/Shared/Models/MediaDirectory.ts
module FC.Shared.Models {
    import CM = FC.Core.CoreModel;
    import INT = FC.Shared.Interfaces;
    import MODELS = FC.Shared.Models;
    import MODULES = FC.Modules;

    export class MediaDirectory  {
        public constructor() {
            this.Media = new Array<FC.Shared.Models.Media>();
            this.Children = new Array<FC.Shared.Models.MediaDirectory>();
        }
        public DirectoryID: string;
        public ParentID: string;
        public Name: string;
        public Media: FC.Shared.Models.Media[];
        public Children: FC.Shared.Models.MediaDirectory[];
        public AuthorID: string;
        public Author: ApplicationUser;
    }
}